s3/pkg/service/object.go

```diff
@@ -116,17 +116,159 @@ func (s *s3Service) PutObject(ctx context.Context, in pb.S3_PutObjectStream) err
 		in.SendMsg(result)
 	}()
 
+	req := &pb.PutObjectRequest{}
+	err = in.RecvMsg(req)
+	if err != nil {
+		log.Errorln("failed to get msg with err:", err)
+		return ErrInternalError
+	}
+
+	log.Infof("*********bucket:%s,key:%s,size:%d\n", req.BucketName, req.ObjectKey, req.Size)
+	bucket, err := s.MetaStorage.GetBucket(ctx, req.BucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	if bucket.Versioning.Status == utils.VersioningEnabled {
+		log.Info("Inside PUT Versioned Objects")
+		return PutVersionedObject(ctx, req, in, s)
+	} else {
+		log.Info("Inside PUT  Objects")
+		return PutObject(ctx, req, in, s)
+	}
+}
+
+func PutVersionedObject(ctx context.Context, req *pb.PutObjectRequest, in pb.S3_PutObjectStream, s *s3Service) error {
```

Comment:
Most part of PutVersionedObject and PutObject are the same, please move those part into a function, and let  PutVersionedObject and PutObject call it.

Reply:
We'll do it in next release